pkg/controller/operators/olm/operatorgroup.go

```diff
@@ -64,102 +66,147 @@ func (a *Operator) syncOperatorGroups(obj interface{}) error {
 		return err
 	}
 	logger.Debug("operatorgroup clusterroles ensured")
 
-	set := a.csvSet(op.Namespace, v1alpha1.CSVPhaseAny)
-	providedAPIs := make(resolver.APISet)
-	for _, csv := range set {
-		logger := logger.WithField("csv", csv.GetName())
+	// Requeue all CSVs that provide the same APIs (including those removed). This notifies conflicting CSVs in
+	// intersecting groups that their conflict has possibly been resolved, either through resizing or through
+	// deletion of the conflicting CSV.
+	groupSurface := resolver.NewOperatorGroup(op)
+	groupProvidedAPIs := groupSurface.ProvidedAPIs()
+	providedAPIsForCSVs := a.providedAPIsFromCSVs(op, logger)
+	providedAPIsForGroup := providedAPIsForCSVs.Union(groupProvidedAPIs)
 
-		// Don't stomp on copied CSVs
-		if csv.IsCopied() {
-			logger.Debug("csv is copied. not updating annotations or including in operatorgroup's provided api set")
-			continue
+	csvs, err := a.findCSVsThatProvideAnyOf(providedAPIsForGroup)
+	if err != nil {
+		logger.WithError(err).Warn("could not find csvs that provide group apis")
+	}
+	for _, csv := range csvs {
+		logger.WithFields(logrus.Fields{
+			"csv":       csv.GetName(),
+			"namespace": csv.GetNamespace(),
+		}).Debug("requeueing provider")
+		if err := a.csvQueueSet.Requeue(csv.GetName(), csv.GetNamespace()); err != nil {
+			logger.WithError(err).Warn("could not requeue provider")
 		}
-		
-		if a.operatorGroupAnnotationsDiffer(&csv.ObjectMeta, op) {
-			a.setOperatorGroupAnnotations(&csv.ObjectMeta, op, true)
+	}
+
+	a.pruneProvidedAPIs(op, groupProvidedAPIs, providedAPIsForCSVs, logger)
+	return nil
+}
+
+func (a *Operator) annotateCSVs(group *v1alpha2.OperatorGroup, targetNamespaces []string, logger *logrus.Entry) error {
+	updateErrs := []error{}
+	targetNamespaceSet := resolver.NewNamespaceSet(targetNamespaces)
+
+	for _, csv := range a.csvSet(group.GetNamespace(), v1alpha1.CSVPhaseAny) {
+		logger := logger.WithField("csv", csv.GetName())
+
+		originalNamespacesAnnotation, _ := a.copyOperatorGroupAnnotations(&csv.ObjectMeta)[v1alpha2.OperatorGroupTargetsAnnotationKey]
+		originalNamespaceSet := resolver.NewNamespaceSetFromString(originalNamespacesAnnotation)
+
+		if a.operatorGroupAnnotationsDiffer(&csv.ObjectMeta, group) {
+			a.setOperatorGroupAnnotations(&csv.ObjectMeta, group, true)
 			// CRDs don't support strategic merge patching, but in the future if they do this should be updated to patch
 			if _, err := a.client.OperatorsV1alpha1().ClusterServiceVersions(csv.GetNamespace()).Update(csv); err != nil && !k8serrors.IsNotFound(err) {
-				// TODO: return an error and requeue the OperatorGroup here? Can this cause an update to never happen if there's resource contention?
 				logger.WithError(err).Warnf("error adding operatorgroup annotations")
-				return err
+				updateErrs = append(updateErrs, err)
+				continue
+			}
+		}
+
+		// requeue csvs in original namespaces or in new target namespaces (to capture removed/added namespaces)
```

Comment:
this ended up being the reason contraction wasn't working after the rebase. somewhere in reconciling those commits we lost the code that requeues CSVs from the _old_ set of namespaces, which is important to make sure we pick up CSVs that need to be cleaned up when an operatorgroup is contracted.

Added some abstractions around namespacesets to help make this clearer

Reply:
Isn't this effectively done around [line 110](https://github.com/operator-framework/operator-lifecycle-manager/pull/736/files#diff-806841494fa68cbd1620970ba03566ccR110)?

Reply:
Line 110 doesn't capture the "contraction" case, right?

Reply:
If you remove the the "not copied" predicate in `getAllCSVsThatProvide(...)`, then I think it would since that returns the set of CSVs across all namespaces that provide the given APIs. As I recall, the input it's given in `syncOperatorGroup` represents the union of APIs in the `providedAPIs` annotation and the CSVs in the `OperatorGroup`'s namespace.